### DISCUSSION TITLE
Refactoring source code

### DISCUSSION THREAD

**Discussion by NTBBloodbath** [rest-nvim/rest.nvim#28] [Ideas]
Hello everyone, I made this discussion in order to discuss something that I have been thinking these last days since I have been unoccupied. It's about refactoring the source code of rest.nvim to make it easier to maintain in the future if the plugin grows more and to divide it so the code will be a lot more readable.

Basically, the main file would be containing only the rest.nvim core; handle rest.nvim buffer & window and the cURL request. Everything related to parsing the http files would go into a parser file that makes more sense and makes the code look cleaner _in my opinion_. It's also intended to make life easier for future rest.nvim contributors _if there are any, I hope so_.

Also btw @disrupted as you are the only contributor at the moment I wanted to know your opinion about it. Lemme know if you have some ideas to improve this so we can implement them too!

**Comment by disrupted** [comment]
hey, thanks for opening this discussion. I think it's a great idea to structure the plugin codebase and make it as simple as possible to add new features moving forward.

I'll have a look to see if I can come up with some ideas and get back to you.

**Reply by NTBBloodbath** [reply]
> I think it's a great idea to structure the plugin codebase and make it as simple as possible to add new features moving forward.

Yes, I think the same thing. Also when I have written the plugin I was still a noob using Lua and now that I have more experience on it I think the code quality can be a lot better now with a rewrite.

> I'll have a look to see if I can come up with some ideas and get back to you.

Sure, thank you for your help!

**Reply by disrupted** [reply]
Sorry it took a while, here are some of my thoughts.

Splitting the plugin code and making it more modular is definitely a good idea. During the rewrite I would also make sure to replace `vim.fn` calls (Vim script) with their `vim.api` counterparts.

Another thing I thought about is Treesitter. I still believe that in an effort to lay some groundwork here it could be really cool to build a Treesitter parser for the `.http` filetype. That way you could remove all the parsing logic of headers, content etc. and use TS queries instead. Also syntax highlighting using TS.
Take this with a grain of salt as I have no experience writing TS grammar but interested to hear your thoughts on this.

**Reply by NTBBloodbath** [reply]
Hey, no worries!

> During the rewrite I would also make sure to replace `vim.fn` calls (Vim script) with their `vim.api` counterparts.

Yeah, we should stop calling VimL chunks to speed up the stuff.

> Another thing I thought about is Treesitter. I still believe that in an effort to lay some groundwork here it could be really cool to build a Treesitter parser for the `.http` filetype.

I was thinking about this too and writing a Treesitter parser for the `http` files should be easy because the syntax is pretty tiny and without complex things. I'll start some tests later, let's hope that we can get rid of the Vim script syntax files in the future.

> That way you could remove all the parsing logic of headers, content etc. and use TS queries instead.

I'm not really advanced with Treesitter stuff, can we use it for this? I mean, as you may know the parsing is for extracting the `.http` file contents so we can pass them to cURL. I'll investigate more about it but sounds great _if it's possible_.
